plonky3/src/circuit_builder.rs

```diff
@@ -6,28 +6,42 @@
 //! everywhere save for at row j is constructed to constrain s * (pub - x) on
 //! every row.
 
-use std::collections::{BTreeMap, HashSet};
+use std::{
+    cell::RefCell,
+    collections::{BTreeMap, BTreeSet, HashMap},
+    sync::Mutex,
+};
 
 use crate::params::{Commitment, FieldElementMap, Plonky3Field, ProverData};
 use p3_air::{Air, AirBuilder, AirBuilderWithPublicValues, BaseAir, PairBuilder};
 use p3_matrix::{dense::RowMajorMatrix, Matrix};
 use powdr_ast::analyzed::{
     AlgebraicBinaryOperation, AlgebraicBinaryOperator, AlgebraicExpression,
-    AlgebraicUnaryOperation, AlgebraicUnaryOperator, Analyzed, Identity, IdentityKind,
-    PolynomialType, SelectedExpressions,
+    AlgebraicUnaryOperation, AlgebraicUnaryOperator, Analyzed, Challenge, Identity, IdentityKind,
+    PolyID, PolynomialType, SelectedExpressions, SymbolKind,
 };
+
+use p3_uni_stark::{CallbackResult, MultiStageAir, MultistageAirBuilder, NextStageTraceCallback};
+use powdr_ast::parsed::visitor::ExpressionVisitable;
+
 use powdr_executor::witgen::WitgenCallback;
-use powdr_number::{DegreeType, FieldElement};
+use powdr_number::FieldElement;
+
+type Witness<T> = Mutex<RefCell<Vec<(String, Vec<T>)>>>;
 
 /// A description of the constraint system.
 /// All of the data is derived from the analyzed PIL, but is materialized
 /// here for performance reasons.
 struct ConstraintSystem<T> {
+    // for each column, the stage (-1 for fixed) and index of this column in this stage
```

Comment:
-1 feels slightly C-like. Don't wanna cause a major refactoring, but is it a lot worse to make it use an enum?

Reply:
Or a separate member for fixed columns. I think we distinguish the two cases anyway wherever this is used.

Reply:
done